### PR TITLE
Regex upgrades

### PR DESCRIPTION
**Pull request by KoenDG**
This very small PR is related to issue #24226 it is small exactly to keep it readable.

This PR is about the regular expressions used in the codebase. Only 2 instances were found where performance optimizations could be recommended. These are both regexes using the `|` symbol to achieve a logical OR comparison. This can be done much more efficiently with a character class.

What is a character class: http://www.regular-expressions.info/charclass.html

Explanation as to why it is faster: http://stackoverflow.com/a/26141949

I made some test cases demonstrating the results are identical.

Version.java: https://ideone.com/hMHGh5

DeprecationLogger.java: https://ideone.com/j9LRU7

===========
1. Contributor agreement: Has been signed
2. Contributor guidelines: I have read them.
3. Gradle check: build successful.
4. PR is against master
5. The PR is not OS specific.
6. I'm not part of a class.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java [hunk 1]
```diff
@@ -311,7 +311,7 @@ public static String formatWarning(final String s) {
      * @return the escaped string
      */
     public static String escape(String s) {
-        return s.replaceAll("(\\\\|\")", "\\\\$1");
+        return s.replaceAll("([\\\\\"])", "\\\\$1");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java]: I think the existing version is more readable and I doubt that the performance difference matters at all.

File: core/src/main/java/org/elasticsearch/Version.java [hunk 2]
```diff
@@ -196,7 +196,7 @@ public static Version fromString(String version) {
         if (snapshot = version.endsWith("-SNAPSHOT")) {
             version = version.substring(0, version.length() - 9);
         }
-        String[] parts = version.split("\\.|\\-");
+        String[] parts = version.split("[.-]");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Version.java]: The performance difference here probably doesn't matter, although the new version is more readable.

**Review by jasontedor — COMMENTED** [review]
I left some comments.

File: core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java]: Maybe we can get both worlds with `[\"\\\\]`; I think that this is as-readable as the current version, and we get the win of the character class change. Can you make this change and I can pull this in?

Comment (KoenDG) [inline core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java]: Updated as suggested.

**Comment by jasontedor** [comment]
ok to test

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by jasontedor** [comment]
Thanks @KoenDG.

**Pull request merged**